### ISSUE TITLE
[bug] conan remove fails when using -p without package id together with -r

### ISSUE THREAD

**Issue by jasal82** [conan-io/conan#7332]
When I do `conan remove -f -p -r <remote> <existing_reference>` with a reference for which there is only a recipe but no binary package on the remote Conan fails with the following error message:

```
ERROR: {
  "errors" : [ {
    "status" : 404,
    "message" : "Couldn't find 'user/pkg/0.1.0/develop/d8cc7d7e1317dcbc71d65614521acffd/package'"
  } ]
}. [Remote: my-remote]
```

I would expect it to succeed even if there is nothing to do.

**Comment by jasal82** [comment]
It seems that the `-p` option without parameters is broken even when there is a binary package on the remote. I still get the same error.

**Comment by jgsogo** [comment]
It totally makes sense. Just to confirm, can you report the server you are using? Artifactory? Which version? Is there any special configuration like virtual repositories?

Thanks!

**Comment by jasal82** [comment]
We're using Artifactory 6.17.0, no virtual repositories, revisions enabled.

**Comment by jasal82** [comment]
I guess when no package ID is specified Conan would have to use the Artifactory API to list the available packages and then iterate on that list to delete them individually. Probably that functionality is currently missing.

**Comment by jasal82** [comment]
Here are some traces

```
{"_action": "COMMAND", "name": "remove", "parameters": {"builds": null, "force": true, "outdated": false, "packages": [], "pattern": "EDP_Assign/1.8.0@sick/develop", "query": null, "remote_name": "edp-conan-local", "src": false}, "time": 1594288541.797875}
{"_action": "REST_API_CALL", "duration": 0.055852413177490234, "headers": {"User-Agent": "Conan/1.27.0 (Python 3.7.2) python-requests/2.21.0", "X-Client-Anonymous-Id": "**********", "X-Client-Id": ""}, "method": "GET", "time": 1594288541.8786561, "url": "https://myartifactory/artifactory/api/conan/edp-conan-local/v1/ping"}
{"_action": "REST_API_CALL", "duration": 0.00894618034362793, "headers": {"User-Agent": "Conan/1.27.0 (Python 3.7.2) python-requests/2.21.0", "X-Client-Anonymous-Id": "**********", "X-Client-Id": ""}, "method": "GET", "time": 1594288541.894613, "url": "https://myartifactory/artifactory/api/conan/edp-conan-local/v2/users/check_credentials"}
{"_action": "REST_API_CALL", "duration": 0.011967897415161133, "headers": {"User-Agent": "Conan/1.27.0 (Python 3.7.2) python-requests/2.21.0", "X-Client-Anonymous-Id": "**********", "X-Client-Id": ""}, "method": "GET", "time": 1594288541.9145598, "url": "https://myartifactory/artifactory/api/conan/edp-conan-local/v2/conans/EDP_Assign/1.8.0/sick/develop/revisions"}
{"_action": "REST_API_CALL", "duration": 0.009973287582397461, "headers": {"User-Agent": "Conan/1.27.0 (Python 3.7.2) python-requests/2.21.0", "X-Client-Anonymous-Id": "**********", "X-Client-Id": ""}, "method": "DELETE", "time": 1594288541.932512, "url": "https://myartifactory/artifactory/api/conan/edp-conan-local/v2/conans/EDP_Assign/1.8.0/sick/develop/revisions/235391/packages"}
{"_action": "EXCEPTION", "class": "NotFoundException", "message": "{\n  \"errors\" : [ {\n    \"status\" : 404,\n    \"message\" : \"Couldn't find 'sick/EDP_Assign/1.8.0/develop/235391/package'\"\n  } ]\n}. [Remote: edp-conan-local]", "time": 1594288541.9424858}
```

**Comment by jasal82** [comment]
There's this function in `conans/server/store/server_store.py` which just unconditionally assumes that there must be a `package` folder:

```python
def remove_all_packages(self, ref):
    assert ref.revision is not None, "BUG: server store needs RREV remove_all_packages"
    assert isinstance(ref, ConanFileReference)
    packages_folder = self.packages(ref)
    self._storage_adapter.delete_folder(packages_folder)
```

**Comment by jgsogo** [comment]
Thanks for the investigation! You are one step away from creating the PR 😉

**Comment by jasal82** [comment]
Hm not sure, that looks like some kind of server side implementation to me because the only storage adapter that I can find is a disk storage adapter. I guess that's not part of the callstack on the client side. Trying to look into this some more but probably I could use some help ;)

**Comment by jasal82** [comment]
Unless that's actually the implementation on Artifactory side. Then a fix would require us to update Artifactory which is not what I want to hear...

**Comment by jgsogo** [comment]
The Conan API with Artifactory is quite straightforward, as you can see from your logs, Artifactory returns an error if the resource you are requesting to DELETE doesn't exist 

```
{"_action": "REST_API_CALL", "duration": 0.009973287582397461, "headers": {"User-Agent": "Conan/1.27.0 (Python 3.7.2) python-requests/2.21.0", "X-Client-Anonymous-Id": "**********", "X-Client-Id": ""}, "method": "DELETE", "time": 1594288541.932512, "url": "https://myartifactory/artifactory/api/conan/edp-conan-local/v2/conans/EDP_Assign/1.8.0/sick/develop/revisions/235391/packages"}
```

This is a design decision in Artifactory that will affect every other package manager and I'm sure it is something we have _to live with_, so we need to implement the **logic in the client-side** (it will take us just one release, it should be really easy):
 a) my first approach would be to check if the message/error returned by Artifactory it is clear about what has happened, Conan can capture that concrete error and ignore it safely.
 b) if that is not possible, Conan can list the packages in the server and, if it is empty, do not call the `remove_all_packages` function.
 c) (Maybe looking at the sources, a `try/catch-pass` around `remove_all_packages` is enough... )

**Comment by memsharded** [comment]
Fix merged, will be in Conan 1.28.
